### ISSUE TITLE
String API Refactor Error Test Case

### ISSUE THREAD

**Issue by Maxwell-Alexius** [goby-lang/goby#184]
According to [Ruby Ver. 2.4.1 Doc](https://ruby-doc.org/core-2.4.1/String.html)

String method implementation listed below:

Implemented and well-documented:
```
String#+
String#*
String#>
String#<
String#<=>
String#!=
String#capitalize
String#upcase
String#downcase
String#size
String#length
String#reverse
String#to_s
String#to_i
```

Haven't implemented yet:

- [ ] `String#<< (Same as #concat)`
- [x] `String#concat`
- [x] `String#[]`
- [x] `String#[]= (Replace the index of the string with another string or character)`
- [ ] `String#%`
- [ ] `String#index`
- [x] `String#to_a`
- [ ] `String#to_sym`
- [x] `String#count`
- [ ] `String#each_byte`
- [ ] `String#each_char`
- [ ] `String#each_line`
- [x] `String#empty?`
- [x] `String#eql?`
- [x] `String#start_with?`
- [x] `String#end_with?`
- [x] `String#insert`
- [x] `String#delete`
- [x] `String#chop (Returns a string which removes last character)`
- [ ] `String#chomp (Returns a new string which default removes separators like carriage returns if present in last of the string)`
- [ ] `String#dump (Returns raw string)`
- [ ] `String#center`
- [x] `String#ljust (String left justify)`
- [x] `String#rjust (String right justify)`
- [ ] `String#lstrip (Left Strip)`
- [ ] `String#rstrip (Right Strip)`
- [x] `String#replace`
- [x] `String#strip`
- [x] `String#slice`
- [x] `String#split`
- [ ] `String#squeeze (e.g. "Hello   I  like    Goby  Lang".squeeze => "Hello I like Goby Lang")`
- [ ] `String#succ`
- [ ] `String#upto`

Some methods from the doc I think is unnecessary or can be renamed?

- [ ] `String#chr (Returns the first character of the string, maybe rename to #start and then we can have the String#last method which calls the last character of the string)`
- [ ] `String#clear (Returns empty string)`

Not sure to implement for now:

- [ ] `String#to_f`
- [ ] `String#to_c`
- [ ] `String#=~`
- [x] `String#gsub`
- [ ] `String#encode`
- [ ] `String#valid_encoding?`

**Comment by st0012** [comment]
@Maxwell-Alexius Can you use `- [ ]` to make list checkable? Then I can just check what I think is needed.

**Comment by Maxwell-Alexius** [comment]
@st0012  Okay, you can check it out now!~

**Comment by st0012** [comment]
@Maxwell-Alexius Checked

**Comment by hachi8833** [comment]
I suppose that `String#include?` is also needed because the implementation would be much simpler than `String#gsub` that requires full regexp library. How about this?

- [ ] `String#include?`

**Comment by Maxwell-Alexius** [comment]
@st0012 This can be closed, I think? Because PR #239 merged.

**Comment by st0012** [comment]
@Maxwell-Alexius Yap
